dpnp/dpnp_iface_mathematical.py

```
@@ -2043,43 +2073,99 @@ def prod(
     >>> import dpnp as np
-    >>> np.prod(np.array([[1, 2], [3, 4]]))
-    24
     >>> np.prod(np.array([1, 2]))
-    2
+    array(2)
+
+    >>> a = np.array([[1, 2], [3, 4]])
+    >>> np.prod(a)
+    array(24)
+
+    >>> np.prod(a, axis=1)
+    array([ 2, 12])
+    >>> np.prod(a, axis=0)
+    array([3, 8])
+
+    >>> x = np.array([1, 2, 3], dtype=np.int8)
+    >>> np.prod(x).dtype == int
+    True
 
     """
 
-    x1_desc = dpnp.get_dpnp_descriptor(x1, copy_when_nondefault_queue=False)
-    if x1_desc:
-        if where is not True:
-            pass
-        else:
-            out_desc = (
-                dpnp.get_dpnp_descriptor(out, copy_when_nondefault_queue=False)
-                if out is not None
-                else None
-            )
-            result_obj = dpnp_prod(
-                x1_desc, axis, dtype, out_desc, keepdims, initial, where
-            ).get_pyobj()
-            result = dpnp.convert_single_elem_array_to_scalar(
-                result_obj, keepdims
-            )
+    # Product reduction for complex output are known to fail for Gen9 with 2024.0 compiler
+    # TODO: get rid of this temporary work around when OneAPI 2024.1 is released
+    if not isinstance(a, (dpnp_array, dpt.usm_ndarray)):
+        raise TypeError(
+            "An array must be any of supported type, but got {}".format(type(a))
+        )
+    _dtypes = (a.dtype, dtype)
+    _any_complex = any(
+        dpt.isdtype(dpnp.dtype(dt), "complex floating") for dt in _dtypes
+    )
+    device_mask = (
+        du.intel_device_info(a.sycl_device).get("device_id", 0) & 0xFF00
+    )
+    if _any_complex and device_mask in [0x3E00, 0x9B00]:
+        return call_origin(
+            numpy.prod,
+            a,
+            axis=axis,
+            dtype=dtype,
+            out=out,
+            keepdims=keepdims,
+            initial=initial,
+            where=where,
+        )
+    if initial is not None:
+        pass
+    elif where is not True:
+        pass
```

can we raise `NotImplementedError` instead of fall back on numpy?